src/sys/unix.rs

```diff
@@ -2154,6 +2157,55 @@ impl crate::Socket {
             )
         }
     }
+
+    /// Get the value of the `TCP_CONGESTION` option for this socket.
+    ///
+    /// For more information about this option, see [`set_tcp_congestion`].
+    ///
+    /// [`set_tcp_congestion`]: Socket::set_tcp_congestion
+    #[cfg(all(feature = "all", any(target_os = "freebsd", target_os = "linux")))]
+    #[cfg_attr(
+        docsrs,
+        doc(cfg(all(feature = "all", any(target_os = "freebsd", target_os = "linux"))))
+    )]
+    pub fn tcp_congestion(&self) -> io::Result<Vec<u8>> {
+        let mut payload: MaybeUninit<[u8; TCP_CA_NAME_MAX]> = MaybeUninit::uninit();
+        let mut len = size_of::<[u8; TCP_CA_NAME_MAX]>() as libc::socklen_t;
+        syscall!(getsockopt(
+            self.as_raw(),
+            IPPROTO_TCP,
+            libc::TCP_CONGESTION,
+            payload.as_mut_ptr().cast(),
+            &mut len,
+        ))
+        .map(|_| {
+            let buf = unsafe { payload.assume_init() };
```

Comment:
This is UB. Only `&buf[0..len]` bytes are initialised by the kernel.

Reply:
This is not resolved. You can't call `payload.assume_init`, it's not **fully** initialised. I think you can simply remove this line and we're good.

Reply:
Not what I asked for, but fine for now.

Allow me to explain the UB in the old code one more time. You create an array of size `TCP_CA_NAME_MAX`, accessing any of those by would be UB. Then you call the kernel to fill the bytes, but it doesn't fill all of them just `n` bytes (return by the system call). With the old code you called `payload.assume_init()`, which requires the entire array all (`TCP_CA_NAME_MAX` bytes) to be initialised, but that's not always the case as it's only up to `n` bytes.

So the only change needed to the old code was to remove the `payload.assume_init()` line. Because the next line creating a slice of only the initialised bytes and the line after that cast it to a slice of bytes (`MaybeUninit<u8>` -> `u8`).